src/core/strvector.c

```diff
@@ -370,60 +346,55 @@ void igraph_strvector_clear(igraph_strvector_t *sv) {
  * smaller, maybe less, depending on memory management.
  */
 
-igraph_error_t igraph_strvector_resize(igraph_strvector_t* v, igraph_integer_t newsize) {
-    igraph_integer_t toadd = newsize - v->len, i, j;
-    char **tmp;
-    igraph_integer_t reallocsize = newsize;
-
-    IGRAPH_ASSERT(v != 0);
-    IGRAPH_ASSERT(v->data != 0);
-    /*   printf("resize %li to %li\n", v->len, newsize); */
-    if (newsize < v->len) {
-        for (i = newsize; i < v->len; i++) {
-            IGRAPH_FREE(v->data[i]);
-        }
-        /* try to give back some space */
-        tmp = IGRAPH_REALLOC(v->data, reallocsize, char*);
-        /*     printf("resize %li to %li, %p\n", v->len, newsize, tmp); */
-        if (tmp != 0) {
-            v->data = tmp;
-        }
-    } else if (newsize > v->len) {
-        igraph_bool_t error = 0;
-        tmp = IGRAPH_REALLOC(v->data, reallocsize, char*);
-        if (tmp == 0) {
-            IGRAPH_ERROR("cannot resize string vector", IGRAPH_ENOMEM); /* LCOV_EXCL_LINE */
+igraph_error_t igraph_strvector_resize(igraph_strvector_t *sv, igraph_integer_t newsize) {
+    igraph_integer_t toadd = newsize - igraph_strvector_size(sv), i, j;
+    igraph_integer_t oldsize = igraph_strvector_size(sv);
+
+    IGRAPH_ASSERT(sv != NULL);
+    IGRAPH_ASSERT(sv->stor_begin != NULL);
+    if (newsize < oldsize) {
+        for (i = newsize; i < oldsize; i++) {
+            IGRAPH_FREE(sv->stor_begin[i]);
         }
-        v->data = tmp;
+        sv->end = sv->stor_begin + newsize;
+    } else if (newsize > oldsize) {
+        IGRAPH_CHECK(igraph_strvector_reserve(sv, newsize));
 
         for (i = 0; i < toadd; i++) {
-            v->data[v->len + i] = IGRAPH_CALLOC(1, char);
-            if (v->data[v->len + i] == 0) {
-                error = 1;
-                break;
-            }
-            v->data[v->len + i][0] = '\0';
-        }
-        if (error) {
-            /* There was an error, free everything we've allocated so far */
-            for (j = 0; j < i; j++) {
-                if (v->data[v->len + i] != 0) {
-                    IGRAPH_FREE(v->data[v->len + i]);
+            sv->stor_begin[oldsize + i] = IGRAPH_CALLOC(1, char);
+            if (sv->stor_begin[oldsize + i] == NULL) {
+                /* LCOV_EXCL_START */
+                for (j = 0; j < i; j++) {
+                    IGRAPH_FREE(sv->stor_begin[oldsize + i]);
```

Comment:
Did you mean the following?

```suggestion
                    IGRAPH_FREE(sv->stor_begin[oldsize + j]);
```